fvm/evm/handler/handler.go

```diff
@@ -0,0 +1,414 @@
+		// this is a fatal error
+		panic(err)
+	}
+	h.newBlockDraft = &types.Block{
+		Height:            lastExecutedBlock.Height + 1,
+		ParentBlockHash:   parentHash,
+		TotalSupply:       lastExecutedBlock.TotalSupply,
+		TransactionHashes: make([]gethCommon.Hash, 0),
+	}
+}
+
+func (h *ContractHandler) updateBlockDraftStateRoot(stateRoot gethCommon.Hash) {
+	if h.newBlockDraft == nil {
+		h.setupNewBlockDraft()
+	}
+	h.newBlockDraft.StateRoot = stateRoot
+}
+
+func (h *ContractHandler) updateBlockDraftTotalSupply(newValue uint64) {
+	if h.newBlockDraft == nil {
+		h.setupNewBlockDraft()
+	}
+	h.newBlockDraft.TotalSupply = newValue
+}
+
+func (h *ContractHandler) getBlockDraftTotalSupply() uint64 {
+	if h.newBlockDraft == nil {
+		h.setupNewBlockDraft()
+	}
+	return h.newBlockDraft.TotalSupply
+}
+
+func (h *ContractHandler) appendTxHashToBlockDraft(txHash gethCommon.Hash) {
+	if h.newBlockDraft == nil {
+		h.setupNewBlockDraft()
+	}
+	h.newBlockDraft.TransactionHashes = append(h.newBlockDraft.TransactionHashes, txHash)
+}
+
+func (h *ContractHandler) getBlockDraftHeight() uint64 {
+	if h.newBlockDraft == nil {
+		h.setupNewBlockDraft()
+	}
+	return h.newBlockDraft.Height
+}
+
+func (h *ContractHandler) commitBlockDraft() {
+	err := h.blockchain.AppendBlock(h.newBlockDraft)
+	handleError(err)
+	h.emitEvent(types.NewBlockExecutedEvent(h.newBlockDraft))
+	// reset it
+	h.newBlockDraft = nil
+}
+
+func (h ContractHandler) checkGasLimit(limit types.GasLimit) {
+	// check gas limit against what has been left on the transaction side
+	if !h.backend.ComputationAvailable(environment.ComputationKindEVMGasUsage, uint(limit)) {
+		handleError(types.ErrInsufficientComputation)
+	}
+}
+
+func (h ContractHandler) meterGasUsage(res *types.Result) {
+	if res != nil {
+		err := h.backend.MeterComputation(environment.ComputationKindEVMGasUsage, uint(res.GasConsumed))
+		handleError(err)
+	}
+}
+
+func (h *ContractHandler) emitEvent(event *types.Event) {
+	// TODO add extra metering for rlp encoding
+	encoded, err := event.Payload.Encode()
+	handleError(err)
+	err = h.backend.EmitFlowEvent(event.Etype, encoded)
+	handleError(err)
+}
+
+func (h *ContractHandler) getBlockContext() types.BlockContext {
+	return types.BlockContext{
+		BlockNumber:            h.getBlockDraftHeight(),
+		DirectCallBaseGasUsage: types.DefaultDirectCallBaseGasUsage,
+	}
+}
+
+type Account struct {
+	isAuthorized bool
+	address      types.Address
+	fch          *ContractHandler
+}
+
+// newAccount creates a new evm account
+func newAccount(fch *ContractHandler, addr types.Address, isAuthorized bool) *Account {
+	return &Account{
+		isAuthorized: isAuthorized,
+		fch:          fch,
+		address:      addr,
+	}
+}
+
+// Address returns the address associated with the bridged account
+func (a *Account) Address() types.Address {
+	return a.address
+}
+
+// Balance returns the balance of this bridged account
+func (a *Account) Balance() types.Balance {
+	ctx := a.fch.getBlockContext()
```

Comment:
should accessing the balance be metered?

Reply:
for read calls, currently we don't meter any computation and only check the storage interaction limits

Reply:
we might want to do it in the future, I'll add a TODO